File: backend/persistence/stats.py
"""
Profil-Statistiken & Highscores (SPEC §5, §34, §35).

Konsequente Fortsetzung des bestehenden Event-Sourcing-Prinzips
(docs/ARCHITEKTUR.md: "State ist immer Replay des Logs"): Statistiken
werden NICHT bei jedem Matchende inkrementell fortgeschrieben (das
waere bei spaeterem Undo/Korrektur eines bereits abgeschlossenen
Matches fehleranfaellig - Doppelzaehlung oder veraltete Werte), sondern
bei jeder Anfrage frisch aus den bereits gespeicherten, abgeschlossenen
Matches berechnet (Replay ueber MatchEngine, dieselbe Quelle wie beim
Fortsetzen eines unterbrochenen Matches). Das ist fuer eine lokale
Trainings-App mit ueberschaubarer Match-Anzahl performant genug und
vermeidet einen kompletten zweiten (Cache-Invalidierungs-)Zustand.

Die vorbereiteten Tabellen `highscores`/`stats_summary` aus dem
urspruenglichen Schema (backend/persistence/db.py) werden deshalb
bewusst NICHT beschrieben - sie blieben ungenutzt, seit sie angelegt
wurden (config_hash war z.B. bislang immer ein leerer String). Diese
Datei berechnet stattdessen alles bei Bedarf direkt aus
matches/match_players/match_events + config_hash (der hier erstmals
tatsaechlich befuellt wird, siehe compute_config_hash()).

gameConfigurationHash (SPEC §34): Hash ueber Spiel-ID + volle
Settings - bewusst ALLE Settings, nicht nur "relevante" (das muesste
pro Spiel einzeln entschieden werden und wuerde stillschweigend
Ergebnisse zusammenlegen, die laut SPEC-Beispiel getrennt bleiben
muessen, z.B. Random Checkout 40-80/6 Darts vs. 80-130/3 Darts).
"""
from __future__ import annotations

import hashlib
import json

from backend.config.games import get_game
from backend.persistence.db import get_connection

# Pro Engine-Familie die EINE Kennzahl, die als "persoenliche
# Bestleistung"/Highscore-Metrik zaehlt (SPEC §34 Beispiel: Bob's 27
# ALL-TIME nach Punktzahl). Muss ein Feld aus dem jeweiligen
# player_state sein (siehe backend/engine/*.py create_player_state).
PRIMARY_METRIC = {
    "x01": "highestCheckout",
    "checkout_range": "highestLevel",
    "catch": "highestLevel",
    "random_checkout": "successfulCheckouts",
    "target_progression": "bestRun",
    "accuracy_progression": "successfulTargets",
    "jdc": "totalScore",
}

# Nur diese Familien haben ein "Checkout" im klassischen Sinn (SPEC
# §35: Checkout %, Average Checkout Darts, Highest Checkout, Scoring
# Average) - Aufnahmen aus anderen Familien (z.B. Around the World)
# wuerden diese Kennzahlen sonst verwaessern.
CHECKOUT_FAMILIES = {"x01", "checkout_range", "catch", "random_checkout"}


def compute_config_hash(game_id: str, settings: dict) -> str:
    canonical = json.dumps({"gameId": game_id, "settings": settings}, sort_keys=True)
    return hashlib.sha256(canonical.encode()).hexdigest()[:12]


def _pct(part: int, whole: int) -> float | None:
    return round(part / whole * 100, 1) if whole else None


def _load_finished_matches(profile_id: str | None = None, game_id: str | None = None) -> list[dict]:
    conn = get_connection()
    query = (
        "SELECT DISTINCT m.id, m.game_id, m.config_json, m.config_hash, "
        "m.finished_at, m.winner_profile_id FROM matches m"
    )
    conditions = ["m.status = 'finished'"]
    params: list[str] = []
    if profile_id is not None:
        query += " JOIN match_players mp ON mp.match_id = m.id"
        conditions.append("mp.profile_id = ?")
        params.append(profile_id)
    if game_id is not None:
        conditions.append("m.game_id = ?")
        params.append(game_id)
    query += " WHERE " + " AND ".join(conditions) + " ORDER BY m.finished_at ASC"
    rows = conn.execute(query, params).fetchall()
    conn.close()
    return [dict(r) for r in rows]


def _replay_match(row: dict):
    from backend.engine.engine import MatchEngine  # spaeter Import: Zyklus vermeiden
    from backend.persistence import matches as matches_db

    game = get_game(row["game_id"])
    if game is None:
        return None
    events = matches_db.load_match_events(row["id"])
    player_ids = matches_db.load_match_player_ids(row["id"])
    players = [{"id": pid, "name": pid} for pid in player_ids]
    try:
        return MatchEngine(row["id"], game, players, json.loads(row["config_json"]), events=events)
    except Exception:
        # Ein einzelnes beschaedigtes/inkompatibles altes Match soll
        # nicht die gesamte Statistik-Abfrage zum Absturz bringen.
        return None


def compute_profile_stats(profile_id: str) -> dict:
    """SPEC §35: langfristige Statistiken ueber alle Spiele hinweg,
    plus Bestleistung je Spiel und eine kompakte Trainingshistorie."""
    rows = _load_finished_matches(profile_id=profile_id)

    games_played = 0
    wins = 0
    total_darts = singles = doubles = triples = bulls = misses = 0
    scoring_points = scoring_visits = 0
    checkouts_completed = checkout_darts_total = highest_checkout = 0
    per_game: dict[str, dict] = {}
    history: list[dict] = []

    for row in rows:
        engine = _replay_match(row)
        if engine is None or profile_id not in engine.player_states:
            continue
        state = engine.player_states[profile_id]
        # Bei Solo-Training ist der einzige Spieler technisch immer
        # "Gewinner" (Engine-Logik fuer den Run-/Match-Abschluss) - das
        # zaehlt hier bewusst NICHT als "Win", sonst waere die Sieg-
        # Quote fuer reines Solo-Training immer 100%.
        won = len(engine.players) > 1 and engine.winner_id == profile_id
        games_played += 1
        wins += 1 if won else 0

        for t in engine.throw_log:
            if t["playerId"] != profile_id:
                continue
            seg = t["segment"]
            number, mult = seg.get("number", 0), seg.get("multiplier", 0)
            total_darts += 1
            if number == 0 or mult == 0:
                misses += 1
            elif number == 25:
                bulls += 1
            elif mult == 1:
                singles += 1
            elif mult == 2:
                doubles += 1
            elif mult == 3:
                triples += 1

        is_checkout_family = engine.family_name in CHECKOUT_FAMILIES
        for v in engine.visit_log:
            if v["playerId"] != profile_id:
                continue
            if not is_checkout_family:
                continue
            scoring_points += v["value"]
            scoring_visits += 1
            if v["outcome"] == "checkout":
                checkouts_completed += 1
                checkout_darts_total += len(v["throws"])
                highest_checkout = max(highest_checkout, v["value"])

        game_id = row["game_id"]
        bucket = per_game.setdefault(game_id, {"gamesPlayed": 0, "wins": 0, "best": None, "metricName": None})
        bucket["gamesPlayed"] += 1
        bucket["wins"] += 1 if won else 0
        metric_name = PRIMARY_METRIC.get(engine.family_name)
        metric_value = state.get(metric_name) if metric_name else None
        bucket["metricName"] = metric_name
        if metric_value is not None and (bucket["best"] is None or metric_value > bucket["best"]):
            bucket["best"] = metric_value

        history.append({
            "matchId": row["id"],
            "gameId": game_id,
            "finishedAt": row["finished_at"],
            "won": won,
            "metricName": metric_name,
            "metricValue": metric_value,
        })

    return {
        "gamesPlayed": games_played,
        "wins": wins,
        "winPercent": _pct(wins, games_played),
        "accuracy": _pct(total_darts - misses, total_darts),
        "singlePercent": _pct(singles, total_darts),
        "doublePercent": _pct(doubles, total_darts),
        "triplePercent": _pct(triples, total_darts),
        "bullPercent": _pct(bulls, total_darts),
        "checkoutPercent": _pct(checkouts_completed, scoring_visits),
        "averageCheckoutDarts": round(checkout_darts_total / checkouts_completed, 2) if checkouts_completed else None,
        "highestCheckout": highest_checkout,
        "scoringAverage": round(scoring_points / scoring_visits, 2) if scoring_visits else None,
        "perGame": per_game,
        "history": list(reversed(history))[:50],
    }


def _format_config_label(game: dict, settings: dict) -> str:
    labels = {f["key"]: f.get("label", f["key"]) for f in game.get("settingsSchema", [])}
    parts = [f"{labels.get(key, key)}: {value}" for key, value in settings.items() if key in labels]
    return ", ".join(parts) if parts else "Standard"


def compute_leaderboards_for_game(game_id: str) -> list[dict]:
    """SPEC §34: ein lokales All-Time-Leaderboard je Spiel - getrennt
    nach Einstellungen (gameConfigurationHash), da unterschiedliche
    Settings nicht vergleichbar sind."""
    game = get_game(game_id)
    if game is None:
        return []
    metric_name = PRIMARY_METRIC.get(game["engineFamily"])
    if metric_name is None:
        return []

    rows = _load_finished_matches(game_id=game_id)
    buckets: dict[str, dict] = {}
    for row in rows:
        engine = _replay_match(row)
        if engine is None:
            continue
        settings = json.loads(row["config_json"])
        config_hash = row["config_hash"] or compute_config_hash(game_id, settings)
        bucket = buckets.setdefault(config_hash, {"configHash": config_hash, "settings": settings, "best": {}})
        for pid, state in engine.player_states.items():
            value = state.get(metric_name)
            if value is None:
                continue
            if pid not in bucket["best"] or value > bucket["best"][pid]:
                bucket["best"][pid] = value

    results = []
    for bucket in buckets.values():
        entries = sorted(bucket["best"].items(), key=lambda kv: kv[1], reverse=True)[:10]
        results.append({
            "configHash": bucket["configHash"],
            "configLabel": _format_config_label(game, bucket["settings"]),
            "metricName": metric_name,
            "entries": [{"profileId": pid, "value": v} for pid, v in entries],
        })
    return results

File: frontend/src/data/stats.ts
// 1:1-Port von backend/persistence/stats.py (Phase E des Client-
// Rewrites, siehe ~/.claude/plans/agile-brewing-wadler.md). Konsequente
// Fortsetzung des Event-Sourcing-Prinzips: Statistiken werden NICHT
// inkrementell fortgeschrieben (fehleranfaellig bei spaeterem Undo/
// Korrektur eines abgeschlossenen Matches), sondern bei jeder Anfrage
// frisch aus den gespeicherten, abgeschlossenen Matches berechnet
// (Replay ueber MatchEngine, dieselbe Quelle wie beim Fortsetzen).
import { GameDefinition, Leaderboard, LeaderboardEntry, ProfileHistoryEntry, ProfileStats } from "../api";
import { MatchEngine, MatchPlayerRef } from "../engine/matchEngine";
import { STATIC_GAMES } from "../staticGames";
import { getProfile } from "./profiles";
import { loadFinishedMatches, loadFinishedMatchesForProfile, StoredMatch } from "./matches";

export { computeConfigHash } from "./configHash";

// Pro Engine-Familie die EINE Kennzahl, die als "persoenliche
// Bestleistung"/Highscore-Metrik zaehlt. Muss ein Feld aus dem
// jeweiligen player_state sein (siehe engine/families/*.ts).
export const PRIMARY_METRIC: Record<string, string> = {
  x01: "highestCheckout",
  checkout_range: "highestLevel",
  catch: "highestLevel",
  random_checkout: "successfulCheckouts",
  target_progression: "bestRun",
  accuracy_progression: "successfulTargets",
  jdc: "totalScore",
  grouping: "totalScore",
};

// Nur diese Familien haben ein "Checkout" im klassischen Sinn
// (Checkout %, Average Checkout Darts, Highest Checkout, Scoring
// Average) - Aufnahmen aus anderen Familien wuerden diese Kennzahlen
// sonst verwaessern.
export const CHECKOUT_FAMILIES = new Set(["x01", "checkout_range", "catch", "random_checkout"]);

function pct(part: number, whole: number): number | null {
  return whole ? Math.round((part / whole) * 1000) / 10 : null;
}

function findGame(gameId: string): GameDefinition | null {
  return STATIC_GAMES.find((g) => g.id === gameId) ?? null;
}

function replayMatch(row: StoredMatch): MatchEngine | null {
  const game = findGame(row.gameId);
  if (!game) return null;
  const players: MatchPlayerRef[] = row.playerIds.map((pid) => ({ id: pid, name: pid }));
  try {
    return new MatchEngine(row.id, game, players, row.settings, row.events);
  } catch {
    // Ein einzelnes beschaedigtes/inkompatibles altes Match soll nicht
    // die gesamte Statistik-Abfrage zum Absturz bringen.
    return null;
  }
}

export async function computeProfileStats(profileId: string): Promise<ProfileStats> {
  const rows = await loadFinishedMatchesForProfile(profileId);

  let gamesPlayed = 0;
  let wins = 0;
  let totalDarts = 0;
  let singles = 0;
  let doubles = 0;
  let triples = 0;
  let bulls = 0;
  let misses = 0;
  let scoringPoints = 0;
  let scoringVisits = 0;
  let checkoutsCompleted = 0;
  let checkoutDartsTotal = 0;
  let highestCheckout = 0;
  const perGame: ProfileStats["perGame"] = {};
  const history: ProfileHistoryEntry[] = [];

  for (const row of rows) {
    const engine = replayMatch(row);
    if (!engine || !(profileId in engine.playerStates)) continue;
    const state = engine.playerStates[profileId];
    // Bei Solo-Training ist der einzige Spieler technisch immer
    // "Gewinner" - das zaehlt bewusst NICHT als "Win", sonst waere die
    // Siegquote fuer reines Solo-Training immer 100%.
    const won = engine.players.length > 1 && engine.winnerId === profileId;
    gamesPlayed += 1;
    if (won) wins += 1;

    for (const t of engine.throwLog) {
      if (t.playerId !== profileId) continue;
      const { number, multiplier } = t.segment;
      totalDarts += 1;
      if (!number || !multiplier) misses += 1;
      else if (number === 25) bulls += 1;
      else if (multiplier === 1) singles += 1;
      else if (multiplier === 2) doubles += 1;
      else if (multiplier === 3) triples += 1;
    }

    const isCheckoutFamily = CHECKOUT_FAMILIES.has(engine.familyName);
    if (isCheckoutFamily) {
      for (const v of engine.visitLog) {
        if (v.playerId !== profileId) continue;
        scoringPoints += v.value;
        scoringVisits += 1;
        if (v.outcome === "checkout") {
          checkoutsCompleted += 1;
          checkoutDartsTotal += v.throws.length;
          highestCheckout = Math.max(highestCheckout, v.value);
        }
      }
    }

    const metricName = PRIMARY_METRIC[engine.familyName] ?? null;
    const metricValue = metricName ? (state[metricName] as number | undefined) ?? null : null;
    const bucket = perGame[row.gameId] ?? { gamesPlayed: 0, wins: 0, best: null, metricName, gameName: findGame(row.gameId)?.name ?? row.gameId };
    bucket.gamesPlayed += 1;
    if (won) bucket.wins += 1;
    bucket.metricName = metricName;
    if (metricValue !== null && (bucket.best === null || metricValue > bucket.best)) bucket.best = metricValue;
    perGame[row.gameId] = bucket;

    history.push({
      matchId: row.id,
      gameId: row.gameId,
      gameName: findGame(row.gameId)?.name ?? row.gameId,
      finishedAt: row.finishedAt ?? "",
      won,
      metricName,
      metricValue,
    });
  }

  return {
    gamesPlayed,
    wins,
    winPercent: pct(wins, gamesPlayed),
    accuracy: pct(totalDarts - misses, totalDarts),
    singlePercent: pct(singles, totalDarts),
    doublePercent: pct(doubles, totalDarts),
    triplePercent: pct(triples, totalDarts),
    bullPercent: pct(bulls, totalDarts),
    checkoutPercent: pct(checkoutsCompleted, scoringVisits),
    averageCheckoutDarts: checkoutsCompleted ? Math.round((checkoutDartsTotal / checkoutsCompleted) * 100) / 100 : null,
    highestCheckout,
    scoringAverage: scoringVisits ? Math.round((scoringPoints / scoringVisits) * 100) / 100 : null,
    perGame,
    history: history.slice(-50).reverse(),
  };
}

function formatConfigLabel(game: GameDefinition, settings: Record<string, unknown>): string {
  const labels = new Map(game.settingsSchema.map((f) => [f.key, f.label ?? f.key]));
  const parts = Object.entries(settings)
    .filter(([key]) => labels.has(key))
    .map(([key, value]) => `${labels.get(key)}: ${value}`);
  return parts.length > 0 ? parts.join(", ") : "Standard";
}

export async function computeLeaderboardsForGame(gameId: string): Promise<Leaderboard[]> {
  const game = findGame(gameId);
  if (!game) return [];
  const metricName = PRIMARY_METRIC[game.engineFamily];
  if (!metricName) return [];

  const rows = await loadFinishedMatches(gameId);
  const buckets = new Map<string, { settings: Record<string, unknown>; best: Map<string, number> }>();

  for (const row of rows) {
    const engine = replayMatch(row);
    if (!engine) continue;
    const configHash = row.configHash;
    const bucket = buckets.get(configHash) ?? { settings: row.settings, best: new Map<string, number>() };
    for (const [pid, state] of Object.entries(engine.playerStates)) {
      const value = state[metricName] as number | undefined;
      if (value === undefined || value === null) continue;
      const existing = bucket.best.get(pid);
      if (existing === undefined || value > existing) bucket.best.set(pid, value);
    }
    buckets.set(configHash, bucket);
  }

  const results: Leaderboard[] = [];
  for (const [configHash, bucket] of buckets.entries()) {
    const ranked = [...bucket.best.entries()].sort((a, b) => b[1] - a[1]).slice(0, 10);
    const entries: LeaderboardEntry[] = [];
    for (const [profileId, value] of ranked) {
      const profile = await getProfile(profileId);
      entries.push({ profileId, name: profile?.name ?? "?", color: profile?.color ?? null, value });
    }
    results.push({ configHash, configLabel: formatConfigLabel(game, bucket.settings), metricName, entries });
  }
  return results;
}
